tools/mlir-gen/MLIRGen.cpp

```diff
@@ -264,6 +309,54 @@ std::string MLIRGenerator::createMetadata() {
   return data;
 }
 
+void MLIRGenerator::computeFlops(ShapedType inputShape,
+                                 ShapedType outputShape) {
+  // Matmul flops = 2 * M * N * K = 2 * prod(inputDims) * N (outShape[1])
+  int64_t mkFlops = 1;
+  for (int i = 0, max = inputShape.getRank(); i < max; i++)
+    mkFlops *= inputShape.getDimSize(i);
+  int outRank = outputShape.getRank();
+  assert((outRank == 2 || outRank == 4) && "Invalid outRank");
+  // Tiled: N = NB * n = outShape[0] + outShape[3]
+  int64_t nFlops = outputShape.getDimSize(outRank - 1);
+  if (outRank > 2)
+    nFlops *= outputShape.getDimSize(1);
+  flops += 2 * mkFlops * nFlops;
+}
+
+Value MLIRGenerator::lowerNamedMatmul(Value input, Value weight, Value output) {
+  auto inputShape = cast<ShapedType>(input.getType());
+  auto weightShape = cast<ShapedType>(weight.getType());
+  auto outputShape = cast<ShapedType>(output.getType());
+  Value namedMatmul;
+  if (inputShape.getRank() == 2) {
+    namedMatmul = builder
+                      .create<linalg::MatmulOp>(
+                          loc, TypeRange{output.getType()},
+                          ValueRange{input, weight}, ValueRange{output})
+                      .getResult(0);
+  } else if (inputShape.getRank() == 4) {
+    SmallVector<OpFoldResult, 4> dims =
+        tensor::getMixedSizes(builder, loc, weight);
+    Value emptyTensor = builder.create<tensor::EmptyOp>(
+        loc, dims, weightShape.getElementType());
```

Comment:
Builder might not complain but I believe you have to swap reorder dimensions to match for the transpose output.